README.md
# massiv-compat

This repository contains packages that make [`massiv`](https://github.com/lehins/massiv)
work with other packages available from Hackage.

| Language | Github Actions |
|:--------:|:--------------:|
| ![GitHub top language](https://img.shields.io/github/languages/top/lehins/massiv-compat.svg) | [![Build Status](https://github.com/lehins/massiv-compat/workflows/massiv-compat-CI/badge.svg)](https://github.com/lehins/massiv-compat/actions)

|    Github Repo     | Hackage | Nightly | LTS | Compatibility with |
|:-------------------|:-------:|:-------:|:---:|:------------------:|
| [`massiv-serialise`](https://github.com/lehins/massiv-compat/tree/master/massiv-serialise)| [![Hackage](https://img.shields.io/hackage/v/massiv-serialise.svg)](https://hackage.haskell.org/package/massiv-serialise)| [![Nightly](https://www.stackage.org/package/massiv-serialise/badge/nightly)](https://www.stackage.org/nightly/package/massiv-serialise)| [![LTS](https://www.stackage.org/package/massiv-serialise/badge/lts)](https://www.stackage.org/lts/package/massiv-serialise)| [`serialise`](https://hackage.haskell.org/package/serialise)
| [`massiv-persist`](https://github.com/lehins/massiv-compat/tree/master/massiv-persist)| [![Hackage](https://img.shields.io/hackage/v/massiv-persist.svg)](https://hackage.haskell.org/package/massiv-persist)| [![Nightly](https://www.stackage.org/package/massiv-persist/badge/nightly)](https://www.stackage.org/nightly/package/massiv-persist)| [![LTS](https://www.stackage.org/package/massiv-persist/badge/lts)](https://www.stackage.org/lts/package/massiv-persist)| [`persist`](https://hackage.haskell.org/package/persist)


massiv-persist/CHANGELOG.md
# Changelog for `massiv-persist`

## 1.0.0.0

* Switch to support `massiv-1.0.0.0`

## 0.1.0.0

* Initial release


massiv-persist/tests/Test/Massiv/PersistSpec.hs
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE AllowAmbiguousTypes #-}

module Test.Massiv.PersistSpec (spec) where

import Data.Persist
import Common
import Massiv.Persist
import Test.Massiv.Core
import Data.Massiv.Array
import Data.Word

roundtrip :: (Eq a, Show a, Persist a) => a -> Property
roundtrip val = either error id (decode (encode val)) === val

roundtripArray ::
     forall r ix e.
     ( Mutable r e
     , Index ix
     , Persist e
     , Persist (Array r ix e)
     , Eq (Array r ix e)
     , Show (Array r ix e)
     )
  => Array r ix e
  -> Property
roundtripArray arr = expectProp $ do
  let arr'' = either error id (runGet getArray (runPut (putArray arr)))
  getComp arr'' `shouldBe` getComp arr
  arr'' `shouldBe` arr
  let arr' = either error id (decode (encode arr))
  getComp arr' `shouldBe` getComp arr
  arr' `shouldBe` arr

roundtripArraySpec ::
     forall r ix e.
     ( Eq (Array r ix e)
     , Show (Array r ix e)
     , Typeable ix
     , Typeable e
     , Arbitrary ix
     , Load r ix e
     , Mutable r e
     , Arbitrary e
     , Persist e
     , Persist (Array r ix e)
     )
  => Spec
roundtripArraySpec =
  prop (showsType @(Array r ix e) "") $ roundtripArray @r @ix @e

spec :: Spec
spec = do
  describe "Persist" $ do
    prop "Comp" $ roundtrip @Comp
    describe "Ix" $ do
      prop "Ix2" $ roundtrip @Ix2
      prop "Ix3" $ roundtrip @Ix3
      prop "Ix4" $ roundtrip @Ix4
      prop "Ix5" $ roundtrip @Ix5
    describe "Sz" $ do
      prop "Sz1" $ roundtrip @Sz1
      prop "Sz2" $ roundtrip @Sz2
      prop "Sz3" $ roundtrip @Sz3
      prop "Sz4" $ roundtrip @Sz4
      prop "Sz5" $ roundtrip @Sz5
    describe "Array" $ do
      describe "P" $ do
        roundtripArraySpec @P @Ix1 @Word8
        roundtripArraySpec @P @Ix2 @Word8
        roundtripArraySpec @P @Ix3 @Word8
        roundtripArraySpec @P @Ix4 @Word8
        roundtripArraySpec @P @Ix5 @Word8
        roundtripArraySpec @P @Ix1 @Word16
        roundtripArraySpec @P @Ix2 @Word16
        roundtripArraySpec @P @Ix3 @Word16
        roundtripArraySpec @P @Ix4 @Word16
        roundtripArraySpec @P @Ix5 @Word16
        roundtripArraySpec @P @Ix1 @Word32
        roundtripArraySpec @P @Ix2 @Word32
        roundtripArraySpec @P @Ix3 @Word32
        roundtripArraySpec @P @Ix4 @Word32
        roundtripArraySpec @P @Ix5 @Word32
        roundtripArraySpec @P @Ix1 @Word64
        roundtripArraySpec @P @Ix2 @Word64
        roundtripArraySpec @P @Ix3 @Word64
        roundtripArraySpec @P @Ix4 @Word64
        roundtripArraySpec @P @Ix5 @Word64
        roundtripArraySpec @P @Ix1 @Word
        roundtripArraySpec @P @Ix2 @Word
        roundtripArraySpec @P @Ix3 @Word
        roundtripArraySpec @P @Ix4 @Word
        roundtripArraySpec @P @Ix5 @Word
      describe "U" $ do
        roundtripArraySpec @U @Ix1 @(Int, Word)
        roundtripArraySpec @U @Ix2 @(Int, Word)
        roundtripArraySpec @U @Ix3 @(Int, Word)
        roundtripArraySpec @U @Ix4 @(Int, Word)
        roundtripArraySpec @U @Ix5 @(Int, Word)
      describe "S" $ do
        roundtripArraySpec @S @Ix1 @Word8
        roundtripArraySpec @S @Ix2 @Word8
        roundtripArraySpec @S @Ix3 @Word8
        roundtripArraySpec @S @Ix4 @Word8
        roundtripArraySpec @S @Ix5 @Word8
        roundtripArraySpec @S @Ix1 @Word16
        roundtripArraySpec @S @Ix2 @Word16
        roundtripArraySpec @S @Ix3 @Word16
        roundtripArraySpec @S @Ix4 @Word16
        roundtripArraySpec @S @Ix5 @Word16
        roundtripArraySpec @S @Ix1 @Word32
        roundtripArraySpec @S @Ix2 @Word32
        roundtripArraySpec @S @Ix3 @Word32
        roundtripArraySpec @S @Ix4 @Word32
        roundtripArraySpec @S @Ix5 @Word32
        roundtripArraySpec @S @Ix1 @Word64
        roundtripArraySpec @S @Ix2 @Word64
        roundtripArraySpec @S @Ix3 @Word64
        roundtripArraySpec @S @Ix4 @Word64
        roundtripArraySpec @S @Ix5 @Word64
        roundtripArraySpec @S @Ix1 @Word
        roundtripArraySpec @S @Ix2 @Word
        roundtripArraySpec @S @Ix3 @Word
        roundtripArraySpec @S @Ix4 @Word
        roundtripArraySpec @S @Ix5 @Word
      describe "B" $ do
        roundtripArraySpec @B @Ix1 @Integer
        roundtripArraySpec @B @Ix2 @Integer
        roundtripArraySpec @B @Ix3 @Integer
        roundtripArraySpec @B @Ix4 @Integer
        roundtripArraySpec @B @Ix5 @Integer
      describe "N" $ do
        roundtripArraySpec @N @Ix1 @Integer
        roundtripArraySpec @N @Ix2 @Integer
        roundtripArraySpec @N @Ix3 @Integer
        roundtripArraySpec @N @Ix4 @Integer
        roundtripArraySpec @N @Ix5 @Integer


massiv-serialise/src/Massiv/Serialise.hs
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -fno-warn-orphans #-}
-- |
-- Module      : Massiv.Serialise
-- Copyright   : (c) Alexey Kuleshevich 2021
-- License     : BSD3
-- Maintainer  : Alexey Kuleshevich <alexey@kuleshevi.ch>
-- Stability   : experimental
-- Portability : non-portable
--
-- This package provides instances for `Serialise` class for all mutable `Array`
-- representations in [@massiv@](https://hackage.haskell.org/package/massiv) package. These
-- instances are provided as orphans from a separate package in order to avoid direct
-- dependency on [@serialise@](https://hackage.haskell.org/package/serialise) package in
-- @massiv@.
--
-- Array serialisation is done by falling back onto instances for `VG.Vector` types from
-- [@vector@](https://hackage.haskell.org/package/vector) package.
--
-- Below is a simple example how to use it. Note a blank module import: @import
-- Massiv.Serialise ()@, which is the only thing needed from this module in order to use
-- provided orphan instances.
--
-- >>> import Massiv.Serialise ()
-- >>> import Data.Massiv.Array as A
-- >>> let arr = A.fromList A.Seq [72,97,115,107,101,108,108] :: A.Vector A.P Int
-- >>> serialise arr
-- "\NUL\a\135\CANH\CANa\CANs\CANk\CANe\CANl\CANl"
-- >>> deserialise (serialise arr) :: A.Vector A.P Int
-- Array P Seq (Sz1 7)
--   [ 72, 97, 115, 107, 101, 108, 108 ]
--
module Massiv.Serialise
  ( -- * Helper functions used to define Serialise instances
    encodeIx
  , decodeIx
  , mkSzFail
  , encodeArray
  , decodeArray
  ) where

import Codec.Serialise
import Codec.Serialise.Decoding
import Codec.Serialise.Encoding
import Control.DeepSeq (NFData)
import Control.Monad
import qualified Control.Monad.Fail as Fail
import Data.Foldable as F
import Data.Massiv.Array
import Data.Massiv.Array.Manifest.Vector
import Data.Proxy
import Data.Typeable
import qualified Data.Vector as V
import qualified Data.Vector.Generic as VG
import qualified Data.Vector.Primitive as VP
import qualified Data.Vector.Storable as VS
import qualified Data.Vector.Unboxed as VU

instance Serialise Comp where
  encode comp =
    case comp of
      Seq      -> encode (0 :: Int)
      ParOn xs -> encode (1 :: Int) <> encode xs
      ParN n   -> encode (2 :: Int) <> encode n
  decode = do
    ty :: Int <- decode
    case ty of
      0 -> pure Seq
      1 -> ParOn <$> decode
      2 -> ParN <$> decode
      n -> Fail.fail $ "Unexpected Comp tag: " <> show n



-- | Encode index
--
-- @since 0.1.0
encodeIx ::
     forall ix. Index ix
  => ix
  -> Encoding
encodeIx = foldlIndex (\ !acc i -> encode i <> acc) mempty

-- | Decode index
--
-- @since 0.1.0
decodeIx ::
     forall s ix. Index ix
  => Decoder s ix
decodeIx = do
  let decodeDim ix dim = do
        i <- decode
        either (Fail.fail . show) pure $! setDimM ix dim i
  F.foldlM decodeDim zeroIndex [1 .. dimensions (Proxy :: Proxy ix)]

instance Serialise Ix2 where
  encode = encodeIx
  decode = decodeIx

instance Index (IxN n) => Serialise (IxN n) where
  encode = encodeIx
  decode = decodeIx



-- | Construct size from index verifying its correctness.
--
-- @since 0.1.0
mkSzFail ::
     forall ix m. (Index ix, Fail.MonadFail m)
  => ix
  -> m (Sz ix)
mkSzFail ix = do
  let guardNegativeOverflow i !acc = do
        when (i < 0) $ Fail.fail $ "Negative size encountered: " <> show i
        let acc' = i * acc
        when (acc' /= 0 && acc' < acc) $ Fail.fail $ "Overflow detected, size is too big: " <> show i
        pure acc'
  Sz ix <$ foldlIndex (\acc i -> acc >>= guardNegativeOverflow i) (pure 1) ix

instance Index ix => Serialise (Sz ix) where
  encode = encodeIx . unSz
  decode = mkSzFail =<< decodeIx



-- | Encode array by using its corresponding vector instance
--
-- @since 0.1.0
encodeArray ::
     forall v r ix e.
     ( Manifest r e
     , Load r ix e
     , Mutable (ARepr v) e
     , VG.Vector v e
     , VRepr (ARepr v) ~ v
     , Serialise (v e)
     )
  => Array r ix e
  -> Encoding
encodeArray arr =
  encode (getComp arr) <> encode (size arr) <> encode (toVector arr :: v e)

-- | Decode array by using its corresponding vector instance
--
-- @since 0.1.0
decodeArray ::
     forall v r ix e s.
     ( Typeable v
     , VG.Vector v e
     , Load r ix e
     , Load (ARepr v) ix e
     , Mutable r e
     , Serialise (v e)
     )
  => Decoder s (Array r ix e)
decodeArray = do
  comp <- decode
  sz <- decode
  vector :: v e <- decode
  either (Fail.fail . show) pure $ fromVectorM comp sz vector

instance (Index ix, Serialise e) => Serialise (Array BL ix e) where
  encode = encodeArray @V.Vector
  decode = decodeArray @V.Vector

instance (Index ix, NFData e, Serialise e) => Serialise (Array B ix e) where
  encode = encode . toLazyArray
  decode = evalLazyArray <$> decode

instance (Index ix, NFData e, Serialise e) => Serialise (Array BN ix e) where
  encode = encode . unwrapNormalForm
  decode = forceLazyArray <$> decode

instance (Index ix, Storable e, Serialise e) => Serialise (Array S ix e) where
  encode = encodeArray @VS.Vector
  decode = decodeArray @VS.Vector

instance (Index ix, Unbox e, Serialise e) => Serialise (Array U ix e) where
  encode = encodeArray @VU.Vector
  decode = decodeArray @VU.Vector

instance (Index ix, Prim e, Serialise e) => Serialise (Array P ix e) where
  encode = encodeArray @VP.Vector
  decode = decodeArray @VP.Vector



massiv-persist/src/Massiv/Persist.hs
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE CPP #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -fno-warn-orphans #-}
-- |
-- Module      : Massiv.Persist
-- Copyright   : (c) Alexey Kuleshevich 2021
-- License     : BSD3
-- Maintainer  : Alexey Kuleshevich <alexey@kuleshevi.ch>
-- Stability   : experimental
-- Portability : non-portable
--
-- This package provides instances for `Persist` class for all mutable `Array`
-- representations in [@massiv@](https://hackage.haskell.org/package/massiv) package. These
-- instances are provided as orphans from a separate package in order to avoid direct
-- dependency on [@persist@](https://hackage.haskell.org/package/persist) package in
-- @massiv@.
--
-- Array elements are serialized in little endian order, which is consistent with the rest
-- of the instances in @persist@ package.
--
-- Below is a simple example how to use it. Note a blank module import: @import
-- Massiv.Persist ()@, which is the only thing needed from this module in order to use
-- provided orphan instances.
--
-- >>> import Massiv.Persist ()
-- >>> import Data.Massiv.Array as A
-- >>> import Data.Int (Int8)
-- >>> let arr = A.fromList A.Seq [72,97,115,107,101,108,108] :: A.Vector A.P Int8
-- >>> encode arr
-- "\NUL\a\NUL\NUL\NUL\NUL\NUL\NUL\NULHaskell"
-- >>> Right arr' = decode (encode arr) :: Either String (A.Vector A.P Int8)
-- >>> arr'
-- Array P Seq (Sz1 7)
--   [ 72, 97, 115, 107, 101, 108, 108 ]
--
module Massiv.Persist
  ( -- * Helper functions used to define Persist instances
    putIx
  , getIx
  , mkSzFail
  , putArray
  , getArray
  , putPrimArray
  , getPrimArray
  , putStorableArray
  , getStorableArray
  ) where

import Control.DeepSeq (NFData)
import Control.Monad
import qualified Control.Monad.Fail as Fail
import qualified Data.ByteString.Internal as BS
import qualified Data.ByteString.Short.Internal as SBS
import Data.Foldable as F
import Data.Massiv.Array as A
import Data.Massiv.Array.Unsafe
import Data.Persist
import Data.Persist.Internal
import qualified Data.Primitive as Primitive
import qualified Data.Primitive.ByteArray as BA
import Data.Proxy
import Data.Word
import Foreign.Ptr
import Foreign.ForeignPtr
import qualified Foreign.Storable as Storable

#include "MachDeps.h"

instance Persist Comp where
  put comp =
    case comp of
      Seq      -> put (0 :: Word8)
      ParOn xs -> put (1 :: Word8) >> put xs
      ParN n   -> put (2 :: Word8) >> put n
  get = do
    ty :: Word8 <- get
    case ty of
      0 -> pure Seq
      1 -> ParOn <$> get
      2 -> ParN <$> get
      n -> Fail.fail $ "Unexpected Comp tag: " <> show n



-- | Encode index
--
-- @since 0.1.0
putIx ::
     forall ix. Index ix
  => ix
  -> Put ()
putIx = foldlIndex (\ !acc i -> put i >> acc) (pure ())

-- | Decode index
--
-- @since 0.1.0
getIx ::
     forall ix. Index ix
  => Get ix
getIx = do
  let getDim ix dim = do
        i <- get
        either (Fail.fail . show) pure $! setDimM ix dim i
  F.foldlM getDim zeroIndex [1 .. dimensions (Proxy :: Proxy ix)]

instance Persist Ix2 where
  put = putIx
  get = getIx

instance Index (IxN n) => Persist (IxN n) where
  put = putIx
  get = getIx



-- | Construct size from index verifying its correctness.
--
-- @since 0.1.0
mkSzFail ::
     forall ix m. (Index ix, Fail.MonadFail m)
  => ix
  -> m (Sz ix)
mkSzFail ix = do
  let guardNegativeOverflow i !acc = do
        when (i < 0) $ Fail.fail $ "Negative size encountered: " <> show i
        let acc' = i * acc
        when (acc' /= 0 && acc' < acc) $ Fail.fail $ "Overflow detected, size is too big: " <> show i
        pure acc'
  Sz ix <$ foldlIndex (\acc i -> acc >>= guardNegativeOverflow i) (pure 1) ix

instance Index ix => Persist (Sz ix) where
  put = putIx . unSz
  get = mkSzFail =<< getIx



-- | Serialize array computation startegy and size
putArrayHeader :: forall r ix e. (Strategy r, Size r, Index ix) => Array r ix e -> Put ()
putArrayHeader arr = do
  put (getComp arr)
  put (size arr)

instance (Index ix, Persist e) => Persist (Array B ix e) where
  put = putArray
  get = getArray

instance (Index ix, NFData e, Persist e) => Persist (Array N ix e) where
  put = putArray
  get = getArray

instance (Index ix, Unbox e, Persist e) => Persist (Array U ix e) where
  put = putArray
  get = getArray

-- | Serialize array as `LittleEndian`
--
-- @since 0.1.0
putArray :: (Manifest r e, Index ix, Persist e) => Array r ix e -> Put ()
putArray arr = do
  putArrayHeader arr
  A.mapM_ put arr

-- | Deserialize array from binary form in `LittleEndian`
--
-- @since 0.1.0
getArray :: (Mutable r e, Index ix, Persist e) => Get (Array r ix e)
getArray = do
  comp <- get
  sz <- get
  setComp comp <$> A.makeArrayA sz (const get)

-- | Serialize primitive array in `LittleEndian` order
--
-- @since 0.1.0
putPrimArray :: forall ix e . (Index ix, Prim e, Persist e) => Array P ix e -> Put ()
putPrimArray arr = do
#ifdef WORDS_BIGENDIAN
  putArray
#else
  putArrayHeader arr
  let eltBytes = Primitive.sizeOf (undefined :: e)
      n = totalElem (size arr) * eltBytes
      _ = put (undefined :: e) -- force `Persist` constraint
      s =
        case unwrapByteArray arr of
          BA.ByteArray ba -> SBS.SBS ba
  grow n
  Put $ \_ p -> do
    SBS.copyToPtr s (unwrapByteArrayOffset arr * eltBytes) p n
    pure $! p `plusPtr` n :!: ()
#endif

-- | Deserialize primitive array in `LittleEndian` order
--
-- @since 0.1.0
getPrimArray :: forall ix e . (Index ix, Prim e, Persist e) => Get (Array P ix e)
getPrimArray = do
#ifdef WORDS_BIGENDIAN
  getArray
#else
  comp <- get
  sz <- get
  let n = totalElem sz * Primitive.sizeOf (undefined :: e)
      _ = put (undefined :: e) -- force `Persist` constraint
  SBS.SBS sbs <- SBS.toShort <$!> getBytes n
  either (Fail.fail . show) pure $
    resizeM sz (fromByteArray comp (BA.ByteArray sbs))
#endif

instance (Index ix, Prim e, Persist e) => Persist (Array P ix e) where
  put = putPrimArray
  get = getPrimArray


-- | Serialize storable array in `LittleEndian` order
--
-- @since 0.1.0
putStorableArray :: forall ix e . (Index ix, Storable e, Persist e) => Array S ix e -> Put ()
putStorableArray arr = do
#ifdef WORDS_BIGENDIAN
  puArray
#else
  let _ = put (undefined :: e) -- force `Persist` constraint
  putArrayHeader arr
  putByteString $!
    case unsafeArrayToForeignPtr arr of
      (fp, len) ->
        BS.PS (castForeignPtr fp) 0 (len * Storable.sizeOf (undefined :: e))
#endif

-- | Deserialize storable array in `LittleEndian` order
--
-- @since 0.1.0
getStorableArray :: forall ix e . (Index ix, Storable e, Persist e) => Get (Array S ix e)
getStorableArray = do
#ifdef WORDS_BIGENDIAN
  getArray
#else
  comp <- get
  sz <- get
  let eltCount = totalElem sz
      eltBytes = Storable.sizeOf (undefined :: e)
      _ = put (undefined :: e) -- force `Persist` constraint
  BS.PS fp off _ <- getByteString (eltCount * eltBytes)
  either (Fail.fail . show) pure $
    resizeM sz $
    unsafeArrayFromForeignPtr0 comp (fp `plusForeignPtr` off) (Sz eltCount)
#endif

instance (Index ix, Storable e, Persist e) => Persist (Array S ix e) where
  put = putStorableArray
  get = getStorableArray
